Sources/SwiftMemcache/MemcachedResponseDecoder.swift

```diff
@@ -0,0 +1,177 @@
+// See CONTRIBUTORS.txt for the list of swift-memcache-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+import NIOPosix
+
+/// Responses look like:
+///
+/// <RC> <datalen*> <flag1> <flag2> <...>\r\n
+///
+/// Where <RC> is a 2 character return code. The number of flags returned are
+/// based off of the flags supplied.
+///
+/// <datalen> is only for responses with payloads, with the return code 'VA'.
+///
+/// Flags are single character codes, ie 'q' or 'k' or 'I', which adjust the
+/// behavior of the command. If a flag requests a response flag (ie 't' for TTL
+/// remaining), it is returned in the same order as they were in the original
+/// command, though this is not strict.
+///
+/// Flags are single character codes, ie 'q' or 'k' or 'O', which adjust the
+/// behavior of a command. Flags may contain token arguments, which come after the
+/// flag and before the next space or newline, ie 'Oopaque' or 'Kuserkey'. Flags
+/// can return new data or reflect information, in the same order they were
+/// supplied in the request. Sending an 't' flag with a get for an item with 20
+/// seconds of TTL remaining, would return 't20' in the response.
+///
+/// All commands accept a tokens 'P' and 'L' which are completely ignored. The
+/// arguments to 'P' and 'L' can be used as hints or path specifications to a
+/// proxy or router inbetween a client and a memcached daemon. For example, a
+/// client may prepend a "path" in the key itself: "mg /path/foo v" or in a proxy
+/// token: "mg foo Lpath/ v" - the proxy may then optionally remove or forward the
+/// token to a memcached daemon, which will ignore them.
+///
+/// Syntax errors are handled the same as noted under 'Error strings' section
+/// below.
+///
+/// For usage examples beyond basic syntax, please see the wiki:
+/// https://github.com/memcached/memcached/wiki/MetaCommands
+struct MemcachedResponseDecoder: NIOSingleStepByteToMessageDecoder {
+    typealias InboundOut = MemcachedResponse
+
+    /// Describes the errors that can occur during the decoding process.
+    enum MemcachedDecoderError: Error {
+        /// This error is thrown when EOF is encountered but there are still
+        /// readable bytes in the buffer, which can indicate a bad message.
+        case unexpectedEOF
+
+        /// This error is thrown when EOF is encountered but the decoder's next step
+        /// is not `.none`. This error suggests that the message ended prematurely,
+        /// possibly indicating a bad message.
+        case unexpectedNextStep(NextStep)
+    }
+
+    /// The next step that the decoder will take. The value of this enum determines how the decoder
+    /// processes the current state of the ByteBuffer.
+    enum NextStep: Hashable {
+        /// No further steps are needed, the decoding process is complete for the current message.
+        case none
```

Comment:
I think we never transition to that state and since our state machine is looping we can get rid of this